### PR TITLE
feat: add PersistentPodState custom workload support

### PR DESCRIPTION
**Pull request by baxiaoshi**
### Ⅰ. Describe what this PR does
PPS(PersistentPodState) currently has support for standard statefulset and kruise's statefulSet. However, some companies will customize workloads(statefulset like) with stable identification pods,  support such workloads, and add support for custom workloads to pps.

### CODE REVIEW

File: docs/proposals/20220824-persistentpodstate-custom-workload-support.md [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+Modify PPS controller logic to support custom statefulset like workload：
+1.add dynamic watch: for watch custom statefulset like workload, automatically create pps if pps persistent annotations are found
+2.add workload finder: used to find workload according to the target of PPS and build ScaleAndSelector object
+### 1.add dynamic watch
+client-go supports dynamic watch, but requires gvr or gvk to provide the resources. There are two ways to get the resources in the cluster that need to be watched：
+
+1）Get the APIResourceList  in the cluster through the ServerPreferredResources interface of the dynamic client. Because we can't distinguish those workloads, we can only watch all of them here, which will have a lot of unnecessary watch overhead.
+2）Find the corresponding workload by the reverse of the pod's ownerReferences. There is a time delay.
+
+It is recommended to use method 2 to get the owner's gvk by reverse lookup and create Informer if there is no watch before.
```
Comment (furykerry) [inline docs/proposals/20220824-persistentpodstate-custom-workload-support.md:32]: PPS rely on the ordinal maintenance of statefulset, and actually pps can works for statefulset-like workload how to ensure custom workload can provide similar capability and avoid incorrect usage? maybe adding a new config in kruise-configuration configmaps

Comment (baxiaoshi) [reply docs/proposals/20220824-persistentpodstate-custom-workload-support.md:35]: Add PPS_Watch_Custom_Workload_WhiteList in kruise-configuration to declare the Workload that needs to be watched

**Review by furykerry — COMMENTED** [review]
/lgtm

File: pkg/controller/util/watch.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	return true
+}
+
+func AddWatcherDynamically(c controller.Controller, h handler.EventHandler, gvk schema.GroupVersionKind) (bool, error) {
+	if !DiscoverGVK(gvk) {
+		klog.Errorf("Failed to find GVK(%v) in cluster", gvk.String())
+		return false, nil
+	}
+
+	if _, ok := watcherMap.Load(gvk); ok {
```
Comment (veophi) [inline pkg/controller/util/watch.go:59]: plz swap this line with `DiscoverGVK`

File: apis/apps/v1alpha1/persistent_pod_state_types.go [hunk 3]
```diff
@@ -44,6 +48,9 @@ type PersistentPodStateSpec struct {
 	// current only support StatefulSet
 	TargetReference TargetReference `json:"targetRef"`
 
+	// Persist the annotations information of the pods that need to be saved
+	PersistentPodAnnotations []string `json:"persistentPodAnnotations,omitempty"`
```
Comment (zmberg) [inline apis/apps/v1alpha1/persistent_pod_state_types.go:52]:
```
// PersistentPodStateSpec defines the desired state of PersistentPodState
type PersistentPodStateSpec struct {
	PersistentPodAnnotations PersistentPodAnnotation `json:"persistentPodAnnotations,omitempty"`
}
type PersistentPodAnnotation struct {
	// By default, original is the targeted key
	OriginalKey string                    `json:"originalKey"`
	TargetedKey string                    `json:"targetKey,omitempty"`
	PatchPolicy PersistentPatchPolicyType `json:"patchPolicy,omitempty"`
}

type PersistentPatchPolicyType string

var (
	PersistentOverwritePatchPolicy PersistentPatchPolicyType = "Overwrite"
	// PersistentMergePatchJsonPatchPolicy PersistentPatchPolicyType = "MergePatchJson"
)
```

File: pkg/controller/persistentpodstate/persistent_pod_state_event_handler.go [hunk 4]
```diff
@@ -41,9 +47,58 @@ var _ handler.EventHandler = &enqueueRequestForPod{}
...
+			if gvk.GroupKind().String() == sgv.String() {
+				enable = true
+				break
+			}
+		}
+		if !enable {
+			continue
+		}
+
+		_, err := util.AddWatcherDynamically(runtimeController, workloadHandler, gvk)
```
Comment (zmberg) [inline pkg/controller/persistentpodstate/persistent_pod_state_event_handler.go:96]: It is recommended not to implement the watch dynamic crd method in the event handler, but in reconcile, please refer to: https://github.com/openkruise/rollouts/blob/master/pkg/controller/batchrelease/ batchrelease_controller.go#L165

File: pkg/controller/persistentpodstate/persistent_pod_state_event_handler.go [hunk 5]
```diff
@@ -254,3 +309,87 @@ func (p *enqueueRequestForKruiseStatefulSet) Update(evt event.UpdateEvent, q wor
...
+	}
+}
+
+func (p *enqueueRequestForStatefulSetLike) Update(evt event.UpdateEvent, q workqueue.RateLimitingInterface) {
+	oWorkload := evt.ObjectOld.(*unstructured.Unstructured)
+	nWorkload := evt.ObjectNew.(*unstructured.Unstructured)
+	od, _ := json.Marshal(oWorkload)
+	nd, _ := json.Marshal(nWorkload)
+	klog.V(3).Infof("enqueueRequestForStatefulSetLike(%s) old workload", oWorkload.GetName(), string(od))
+	klog.V(3).Infof("enqueueRequestForStatefulSetLike(%s) new workload", oWorkload.GetName(), string(nd))
```
Comment (zmberg) [inline pkg/controller/persistentpodstate/persistent_pod_state_event_handler.go:335]: These two logs remove it, it is estimated that will brush the screen.

File: pkg/controller/persistentpodstate/persistent_pod_state_event_handler.go [hunk 6]
```diff
@@ -254,3 +309,87 @@ func (p *enqueueRequestForKruiseStatefulSet) Update(evt event.UpdateEvent, q wor
...
+			q.Add(reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name:      pps.Name,
+					Namespace: pps.Namespace,
+				},
+			})
+		}
+	}
+
+	if nAnnotations[appsv1alpha1.AnnotationAutoGeneratePersistentPodState] == "true" &&
```
Comment (zmberg) [inline pkg/controller/persistentpodstate/persistent_pod_state_event_handler.go:360]: Can this be removed?

File: pkg/util/configuration/configuration.go [hunk 7]
```diff
@@ -20,6 +20,8 @@ import (
 	"context"
 	"encoding/json"
 
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
```
Comment (zmberg) [inline pkg/util/configuration/configuration.go:24]: go imports -ed

File: pkg/webhook/persistentpodstate/validating/persistent_create_update_handler.go [hunk 8]
```diff
@@ -95,19 +105,17 @@ func validateUpdateObjImmutable(obj, old *appsv1alpha1.PersistentPodState, fldPa
...
 	}
-	_, err := schema.ParseGroupVersion(spec.TargetReference.APIVersion)
-	if err != nil {
-		allErrs = append(allErrs, field.Invalid(fldPath.Child("TargetReference"), spec.TargetReference, err.Error()))
-	}
-	if spec.TargetReference.Kind != "StatefulSet" {
-		allErrs = append(allErrs, field.Invalid(fldPath.Child("TargetReference"), spec.TargetReference, "TargetReference.Kind must be StatefulSet"))
+
+	apiVersion, kind := spec.TargetReference.APIVersion, spec.TargetReference.Kind
+	if spec.TargetReference.Kind != "StatefulSet" && !whiteList.ValidateAPIVersionAndKind(apiVersion, kind) {
```
Comment (zmberg) [inline pkg/webhook/persistentpodstate/validating/persistent_create_update_handler.go:117]: I think not the native or kruise StatefulSet, all other CRD(e.g. other StatefulSet) should judge whiteList.

File: pkg/webhook/persistentpodstate/validating/persistent_validating_test.go [hunk 9]
```diff
@@ -20,6 +20,8 @@ import (
 	"context"
 	"testing"
 
+	corev1 "k8s.io/api/core/v1"
+
```
Comment (zmberg) [inline pkg/webhook/persistentpodstate/validating/persistent_validating_test.go:24]: go imports -ed

File: pkg/webhook/pod/mutating/persistent_pod_state.go [hunk 10]
```diff
@@ -41,8 +42,13 @@ func (h *PodCreateHandler) persistentPodStateMutatingPod(ctx context.Context, re
...
 		return true, nil
 	}
+
+	whiteList, err := configuration.GetPPSWatchWatchCustomWorkloadWhiteList(h.Client)
+	if err != nil {
+		return false, err
+	}
 	ref := metav1.GetControllerOf(pod)
-	if ref == nil || ref.Kind != "StatefulSet" {
+	if ref == nil || (ref.Kind != "StatefulSet" && !whiteList.ValidateAPIVersionAndKind(ref.APIVersion, ref.Kind)) {
```
Comment (zmberg) [inline pkg/webhook/pod/mutating/persistent_pod_state.go:51]: If Custom Kind StatefulSet but not in whitelist, i think should skip the webhook.

File: pkg/controller/persistentpodstate/persistent_pod_state_controller.go [hunk 11]
```diff
@@ -23,6 +23,10 @@ import (
 	"strconv"
 	"strings"
 
+	"github.com/openkruise/kruise/pkg/util/configuration"
+
```
Comment (veophi) [inline pkg/controller/persistentpodstate/persistent_pod_state_controller.go:27]: plz format import

File: pkg/controller/persistentpodstate/persistent_pod_state_event_handler.go [hunk 12]
```diff
@@ -20,16 +20,20 @@ import (
 	"context"
 	"fmt"
 
+	"github.com/openkruise/kruise/pkg/util/configuration"
+
+	"k8s.io/klog/v2"
```
Comment (veophi) [inline pkg/controller/persistentpodstate/persistent_pod_state_event_handler.go:25]: plz format imports

File: pkg/controller/util/watch.go [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+var watcherMap = sync.Map{}
+
+func DiscoverGVK(gvk schema.GroupVersionKind) bool {
```
Comment (veophi) [inline pkg/controller/util/watch.go:20]: this func has been defined in `[pkg/util/discovery/discovery.go](https://github.com/openkruise/kruise/blob/14e21f5bcfd40790861e6db3dc75655c9c37b189/pkg/util/discovery/discovery.go)`

File: pkg/util/configuration/configuration.go [hunk 14]
```diff
@@ -49,12 +51,30 @@ func GetSidecarSetPatchMetadataWhiteList(client client.Client) (*SidecarSetPatch
 	return whiteList, nil
 }
 
+func GetPPSWatchWatchCustomWorkloadWhiteList(client client.Client) (*PPSWatchWatchCustomWorkloadWhiteList, error) {
+	whiteList := &PPSWatchWatchCustomWorkloadWhiteList{Workloads: make([]schema.GroupVersionKind, 0)}
+	data, err := getKruiseConfiguration(client)
```
Comment (veophi) [inline pkg/util/configuration/configuration.go:56]: Maybe, we can just initialize the whiteList in a `init` function at the moment of  kruise starting?

**Comment by veophi** [comment]
Very awesome work, thanks!

**Pull request merged**